File: app/src/main/java/eu/kanade/mangafeed/ui/manga/chapter/ChaptersFragment.java
package eu.kanade.mangafeed.ui.manga.chapter;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import eu.kanade.mangafeed.R;
import eu.kanade.mangafeed.data.database.models.Chapter;
import eu.kanade.mangafeed.data.download.DownloadService;
import eu.kanade.mangafeed.data.download.model.Download;
import eu.kanade.mangafeed.ui.base.activity.BaseActivity;
import eu.kanade.mangafeed.ui.base.adapter.FlexibleViewHolder;
import eu.kanade.mangafeed.ui.base.fragment.BaseRxFragment;
import eu.kanade.mangafeed.ui.decoration.DividerItemDecoration;
import eu.kanade.mangafeed.ui.manga.MangaActivity;
import eu.kanade.mangafeed.ui.reader.ReaderActivity;
import eu.kanade.mangafeed.util.ToastUtil;
import nucleus.factory.RequiresPresenter;
import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

@RequiresPresenter(ChaptersPresenter.class)
public class ChaptersFragment extends BaseRxFragment<ChaptersPresenter> implements
        ActionMode.Callback, FlexibleViewHolder.OnListItemClickListener {

    @Bind(R.id.chapter_list) RecyclerView recyclerView;
    @Bind(R.id.swipe_refresh) SwipeRefreshLayout swipeRefresh;
    @Bind(R.id.toolbar_bottom) ViewGroup toolbarBottom;

    @Bind(R.id.action_sort) ImageView sortBtn;
    @Bind(R.id.action_next_unread) ImageView nextUnreadBtn;
    @Bind(R.id.action_show_unread) CheckBox readCb;
    @Bind(R.id.action_show_downloaded) CheckBox downloadedCb;

    private ChaptersAdapter adapter;
    private LinearLayoutManager linearLayout;
    private ActionMode actionMode;

    private Subscription downloadProgressSubscription;

    public static ChaptersFragment newInstance() {
        return new ChaptersFragment();
    }

    @Override
    public void onCreate(Bundle savedState) {
        super.onCreate(savedState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_manga_chapters, container, false);
        ButterKnife.bind(this, view);

        // Init RecyclerView and adapter
        linearLayout = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(linearLayout);
        recyclerView.addItemDecoration(new DividerItemDecoration(ContextCompat.getDrawable(getContext(), R.drawable.line_divider)));
        recyclerView.setHasFixedSize(true);
        adapter = new ChaptersAdapter(this);
        recyclerView.setAdapter(adapter);

        // Set initial values
        setReadFilter();
        setDownloadedFilter();
        setSortIcon();

        // Init listeners
        swipeRefresh.setOnRefreshListener(this::fetchChapters);
        readCb.setOnCheckedChangeListener((arg, isChecked) ->
                getPresenter().setReadFilter(isChecked));
        downloadedCb.setOnCheckedChangeListener((v, isChecked) ->
                getPresenter().setDownloadedFilter(isChecked));
        sortBtn.setOnClickListener(v -> {
            getPresenter().revertSortOrder();
            setSortIcon();
        });
        nextUnreadBtn.setOnClickListener(v -> {
            Chapter chapter = getPresenter().getNextUnreadChapter();
            if (chapter != null) {
                openChapter(chapter);
            } else {
                ToastUtil.showShort(getContext(), R.string.no_next_chapter);
            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        observeChapterDownloadProgress();
    }

    @Override
    public void onPause() {
        unsubscribeChapterDownloadProgress();
        super.onPause();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.chapters, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_refresh:
                fetchChapters();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    public void onNextChapters(List<Chapter> chapters) {
        // If the list is empty, fetch chapters from source if the conditions are met
        // We use presenter chapters instead because they are always unfiltered
        if (getPresenter().getChapters().isEmpty())
            initialFetchChapters();

        destroyActionModeIfNeeded();
        adapter.setItems(chapters);
    }

    private void initialFetchChapters() {
        // Only fetch if this view is from the catalog and it hasn't requested previously
        if (isCatalogueManga() && !getPresenter().hasRequested()) {
            fetchChapters();
        }
    }

    public void fetchChapters() {
        swipeRefresh.setRefreshing(true);
        getPresenter().fetchChaptersFromSource();
    }

    public void onFetchChaptersDone() {
        swipeRefresh.setRefreshing(false);
    }

    public void onFetchChaptersError() {
        swipeRefresh.setRefreshing(false);
        ToastUtil.showShort(getContext(), R.string.fetch_chapters_error);
    }

    public boolean isCatalogueManga() {
        return ((MangaActivity) getActivity()).isCatalogueManga();
    }

    protected void openChapter(Chapter chapter) {
        getPresenter().onOpenChapter(chapter);
        Intent intent = ReaderActivity.newIntent(getActivity());
        startActivity(intent);
    }

    private void observeChapterDownloadProgress() {
        downloadProgressSubscription = getPresenter().getDownloadProgressObs()
                .subscribe(this::onDownloadProgressChange,
                        error -> { /* TODO getting a NPE sometimes on 'manga' from presenter */ });
    }

    private void unsubscribeChapterDownloadProgress() {
        if (downloadProgressSubscription != null)
            downloadProgressSubscription.unsubscribe();
    }

    private void onDownloadProgressChange(Download download) {
        ChaptersHolder holder = getHolder(download.chapter);
        if (holder != null)
            holder.onProgressChange(getContext(), download.downloadedImages, download.pages.size());
    }

    public void onChapterStatusChange(Chapter chapter) {
        ChaptersHolder holder = getHolder(chapter);
        if (holder != null)
            holder.onStatusChange(chapter.status);
    }

    @Nullable
    private ChaptersHolder getHolder(Chapter chapter) {
        return (ChaptersHolder) recyclerView.findViewHolderForItemId(chapter.id);
    }

    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
        mode.getMenuInflater().inflate(R.menu.chapter_selection, menu);
        adapter.setMode(ChaptersAdapter.MODE_MULTI);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
        return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_select_all:
                return onSelectAll();
            case R.id.action_mark_as_read:
                return onMarkAsRead(getSelectedChapters());
            case R.id.action_mark_as_unread:
                return onMarkAsUnread(getSelectedChapters());
            case R.id.action_download:
                return onDownload(getSelectedChapters());
            case R.id.action_delete:
                return onDelete(getSelectedChapters());
        }
        return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
        adapter.setMode(ChaptersAdapter.MODE_SINGLE);
        adapter.clearSelection();
        actionMode = null;
    }

    private Observable<Chapter> getSelectedChapters() {
        // Create a blocking copy of the selected chapters.
        // When the action mode is closed the list is cleared. If we use background
        // threads with this observable, some emissions could be lost.
        List<Chapter> chapters = Observable.from(adapter.getSelectedItems())
                .map(adapter::getItem).toList().toBlocking().single();

        return Observable.from(chapters);
    }

    public void destroyActionModeIfNeeded() {
        if (actionMode != null) {
            actionMode.finish();
        }
    }

    protected boolean onSelectAll() {
        adapter.selectAll();
        setContextTitle(adapter.getSelectedItemCount());
        return true;
    }

    protected boolean onMarkAsRead(Observable<Chapter> chapters) {
        getPresenter().markChaptersRead(chapters, true);
        return true;
    }

    protected boolean onMarkAsUnread(Observable<Chapter> chapters) {
        getPresenter().markChaptersRead(chapters, false);
        return true;
    }

    protected boolean onDownload(Observable<Chapter> chapters) {
        DownloadService.start(getActivity());

        Observable<Chapter> observable = chapters
                .doOnCompleted(adapter::notifyDataSetChanged);

        getPresenter().downloadChapters(observable);
        destroyActionModeIfNeeded();
        return true;
    }

    protected boolean onDelete(Observable<Chapter> chapters) {
        int size = adapter.getSelectedItemCount();

        MaterialDialog dialog = new MaterialDialog.Builder(getActivity())
                .title(R.string.deleting)
                .progress(false, size, true)
                .cancelable(false)
                .show();

        Observable<Chapter> observable = chapters
                .concatMap(chapter -> {
                    getPresenter().deleteChapter(chapter);
                    return Observable.just(chapter);
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnNext(chapter -> {
                    dialog.incrementProgress(1);
                    chapter.status = Download.NOT_DOWNLOADED;
                })
                .doOnCompleted(adapter::notifyDataSetChanged)
                .finallyDo(dialog::dismiss);

        getPresenter().deleteChapters(observable);
        destroyActionModeIfNeeded();
        return true;
    }

    @Override
    public boolean onListItemClick(int position) {
        if (actionMode != null && adapter.getMode() == ChaptersAdapter.MODE_MULTI) {
            toggleSelection(position);
            return true;
        } else {
            openChapter(adapter.getItem(position));
            return false;
        }
    }

    @Override
    public void onListItemLongClick(int position) {
        if (actionMode == null)
            actionMode = ((BaseActivity) getActivity()).startSupportActionMode(this);

        toggleSelection(position);
    }

    private void toggleSelection(int position) {
        adapter.toggleSelection(position, false);

        int count = adapter.getSelectedItemCount();
        if (count == 0) {
            actionMode.finish();
        } else {
            setContextTitle(count);
            actionMode.invalidate();
        }
    }

    private void setContextTitle(int count) {
        actionMode.setTitle(getString(R.string.label_selected, count));
    }

    public void setSortIcon() {
        if (sortBtn != null) {
            boolean aToZ = getPresenter().getSortOrder();
            sortBtn.setImageResource(!aToZ ? R.drawable.ic_expand_less_white_36dp : R.drawable.ic_expand_more_white_36dp);
        }
    }

    public void setReadFilter() {
        if (readCb != null) {
            readCb.setChecked(getPresenter().getReadFilter());
        }
    }

    public void setDownloadedFilter() {
        if (downloadedCb != null) {
            downloadedCb.setChecked(getPresenter().getDownloadedFilter());
        }
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/chapter/ChaptersFragment.kt
package eu.kanade.tachiyomi.ui.manga.chapter

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.DialogFragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.view.ActionMode
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.*
import com.afollestad.materialdialogs.MaterialDialog
import eu.davidea.flexibleadapter.FlexibleAdapter
import eu.kanade.tachiyomi.R
import eu.kanade.tachiyomi.data.database.models.Chapter
import eu.kanade.tachiyomi.data.database.models.Manga
import eu.kanade.tachiyomi.data.download.model.Download
import eu.kanade.tachiyomi.ui.base.fragment.BaseRxFragment
import eu.kanade.tachiyomi.ui.manga.MangaActivity
import eu.kanade.tachiyomi.ui.reader.ReaderActivity
import eu.kanade.tachiyomi.util.getCoordinates
import eu.kanade.tachiyomi.util.snack
import eu.kanade.tachiyomi.util.toast
import eu.kanade.tachiyomi.widget.DeletingChaptersDialog
import kotlinx.android.synthetic.main.fragment_manga_chapters.*
import nucleus.factory.RequiresPresenter
import timber.log.Timber

@RequiresPresenter(ChaptersPresenter::class)
class ChaptersFragment : BaseRxFragment<ChaptersPresenter>(),
        ActionMode.Callback,
        FlexibleAdapter.OnItemClickListener,
        FlexibleAdapter.OnItemLongClickListener {

    companion object {
        /**
         * Creates a new instance of this fragment.
         *
         * @return a new instance of [ChaptersFragment].
         */
        fun newInstance(): ChaptersFragment {
            return ChaptersFragment()
        }

    }

    /**
     * Adapter containing a list of chapters.
     */
    private lateinit var adapter: ChaptersAdapter

    /**
     * Action mode for multiple selection.
     */
    private var actionMode: ActionMode? = null

    override fun onCreate(savedState: Bundle?) {
        super.onCreate(savedState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_manga_chapters, container, false)
    }

    override fun onViewCreated(view: View, savedState: Bundle?) {
        // Init RecyclerView and adapter
        adapter = ChaptersAdapter(this)

        recycler.adapter = adapter
        recycler.layoutManager = LinearLayoutManager(activity)
        recycler.addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
        recycler.setHasFixedSize(true)
//        TODO enable in a future commit
//        adapter.setFastScroller(fast_scroller, context.getResourceColor(R.attr.colorAccent))
//        adapter.toggleFastScroller()

        swipe_refresh.setOnRefreshListener { fetchChapters() }

        fab.setOnClickListener {
            val item = presenter.getNextUnreadChapter()
            if (item != null) {
                // Create animation listener
                val revealAnimationListener: Animator.AnimatorListener = object : AnimatorListenerAdapter() {
                    override fun onAnimationStart(animation: Animator?) {
                        openChapter(item.chapter, true)
                    }
                }

                // Get coordinates and start animation
                val coordinates = fab.getCoordinates()
                if (!reveal_view.showRevealEffect(coordinates.x, coordinates.y, revealAnimationListener)) {
                    openChapter(item.chapter)
                }
            } else {
                context.toast(R.string.no_next_chapter)
            }
        }
    }

    override fun onResume() {
        // Check if animation view is visible
        if (reveal_view.visibility == View.VISIBLE) {
            // Show the unReveal effect
            val coordinates = fab.getCoordinates()
            reveal_view.hideRevealEffect(coordinates.x, coordinates.y, 1920)
        }
        super.onResume()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.chapters, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        // Initialize menu items.
        val menuFilterRead = menu.findItem(R.id.action_filter_read) ?: return
        val menuFilterUnread = menu.findItem(R.id.action_filter_unread)
        val menuFilterDownloaded = menu.findItem(R.id.action_filter_downloaded)
        val menuFilterBookmarked = menu.findItem(R.id.action_filter_bookmarked)

        // Set correct checkbox values.
        menuFilterRead.isChecked = presenter.onlyRead()
        menuFilterUnread.isChecked = presenter.onlyUnread()
        menuFilterDownloaded.isChecked = presenter.onlyDownloaded()
        menuFilterBookmarked.isChecked = presenter.onlyBookmarked()

        if (presenter.onlyRead())
            //Disable unread filter option if read filter is enabled.
            menuFilterUnread.isEnabled = false
        if (presenter.onlyUnread())
            //Disable read filter option if unread filter is enabled.
            menuFilterRead.isEnabled = false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_display_mode -> showDisplayModeDialog()
            R.id.manga_download -> showDownloadDialog()
            R.id.action_sorting_mode -> showSortingDialog()
            R.id.action_filter_unread -> {
                item.isChecked = !item.isChecked
                presenter.setUnreadFilter(item.isChecked)
                activity.supportInvalidateOptionsMenu()
            }
            R.id.action_filter_read -> {
                item.isChecked = !item.isChecked
                presenter.setReadFilter(item.isChecked)
                activity.supportInvalidateOptionsMenu()
            }
            R.id.action_filter_downloaded -> {
                item.isChecked = !item.isChecked
                presenter.setDownloadedFilter(item.isChecked)
            }
            R.id.action_filter_bookmarked -> {
                item.isChecked = !item.isChecked
                presenter.setBookmarkedFilter(item.isChecked)
            }
            R.id.action_filter_empty -> {
                presenter.removeFilters()
                activity.supportInvalidateOptionsMenu()
            }
            R.id.action_sort -> presenter.revertSortOrder()
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    @Suppress("UNUSED_PARAMETER")
    fun onNextManga(manga: Manga) {
        // Set initial values
        activity.supportInvalidateOptionsMenu()
    }

    fun onNextChapters(chapters: List<ChapterItem>) {
        // If the list is empty, fetch chapters from source if the conditions are met
        // We use presenter chapters instead because they are always unfiltered
        if (presenter.chapters.isEmpty())
            initialFetchChapters()

        destroyActionModeIfNeeded()
        adapter.updateDataSet(chapters)
    }

    private fun initialFetchChapters() {
        // Only fetch if this view is from the catalog and it hasn't requested previously
        if (isCatalogueManga && !presenter.hasRequested) {
            fetchChapters()
        }
    }

    fun fetchChapters() {
        swipe_refresh.isRefreshing = true
        presenter.fetchChaptersFromSource()
    }

    fun onFetchChaptersDone() {
        swipe_refresh.isRefreshing = false
    }

    fun onFetchChaptersError(error: Throwable) {
        swipe_refresh.isRefreshing = false
        context.toast(error.message)
    }

    val isCatalogueManga: Boolean
        get() = (activity as MangaActivity).fromCatalogue

    fun openChapter(chapter: Chapter, hasAnimation: Boolean = false) {
        val intent = ReaderActivity.newIntent(activity, presenter.manga, chapter)
        if (hasAnimation) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
        }
        startActivity(intent)
    }

    private fun showDisplayModeDialog() {
        // Get available modes, ids and the selected mode
        val modes = intArrayOf(R.string.show_title, R.string.show_chapter_number)
        val ids = intArrayOf(Manga.DISPLAY_NAME, Manga.DISPLAY_NUMBER)
        val selectedIndex = if (presenter.manga.displayMode == Manga.DISPLAY_NAME) 0 else 1

        MaterialDialog.Builder(activity)
                .title(R.string.action_display_mode)
                .items(modes.map { getString(it) })
                .itemsIds(ids)
                .itemsCallbackSingleChoice(selectedIndex) { _, itemView, _, _ ->
                    // Save the new display mode
                    presenter.setDisplayMode(itemView.id)
                    // Refresh ui
                    adapter.notifyItemRangeChanged(0, adapter.itemCount)
                    true
                }
                .show()
    }

    private fun showSortingDialog() {
        // Get available modes, ids and the selected mode
        val modes = intArrayOf(R.string.sort_by_source, R.string.sort_by_number)
        val ids = intArrayOf(Manga.SORTING_SOURCE, Manga.SORTING_NUMBER)
        val selectedIndex = if (presenter.manga.sorting == Manga.SORTING_SOURCE) 0 else 1

        MaterialDialog.Builder(activity)
                .title(R.string.sorting_mode)
                .items(modes.map { getString(it) })
                .itemsIds(ids)
                .itemsCallbackSingleChoice(selectedIndex) { _, itemView, _, _ ->
                    // Save the new sorting mode
                    presenter.setSorting(itemView.id)
                    true
                }
                .show()
    }

    private fun showDownloadDialog() {
        // Get available modes
        val modes = intArrayOf(R.string.download_1, R.string.download_5, R.string.download_10,
                R.string.download_unread, R.string.download_all)

        MaterialDialog.Builder(activity)
                .title(R.string.manga_download)
                .negativeText(android.R.string.cancel)
                .items(modes.map { getString(it) })
                .itemsCallback { _, _, i, _ ->

                    fun getUnreadChaptersSorted() = presenter.chapters
                            .filter { !it.read && it.status == Download.NOT_DOWNLOADED }
                            .distinctBy { it.name }
                            .sortedByDescending { it.source_order }

                    // i = 0: Download 1
                    // i = 1: Download 5
                    // i = 2: Download 10
                    // i = 3: Download unread
                    // i = 4: Download all
                    val chaptersToDownload = when (i) {
                        0 -> getUnreadChaptersSorted().take(1)
                        1 -> getUnreadChaptersSorted().take(5)
                        2 -> getUnreadChaptersSorted().take(10)
                        3 -> presenter.chapters.filter { !it.read }
                        4 -> presenter.chapters
                        else -> emptyList()
                    }

                    if (chaptersToDownload.isNotEmpty()) {
                        downloadChapters(chaptersToDownload)
                    }
                }
                .show()
    }

    fun onChapterStatusChange(download: Download) {
        getHolder(download.chapter)?.notifyStatus(download.status)
    }

    private fun getHolder(chapter: Chapter): ChapterHolder? {
        return recycler.findViewHolderForItemId(chapter.id!!) as? ChapterHolder
    }

    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
        mode.menuInflater.inflate(R.menu.chapter_selection, menu)
        adapter.mode = FlexibleAdapter.MODE_MULTI
        return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
        return false
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_select_all -> selectAll()
            R.id.action_mark_as_read -> markAsRead(getSelectedChapters())
            R.id.action_mark_as_unread -> markAsUnread(getSelectedChapters())
            R.id.action_download -> downloadChapters(getSelectedChapters())
            R.id.action_delete -> {
                MaterialDialog.Builder(activity)
                        .content(R.string.confirm_delete_chapters)
                        .positiveText(android.R.string.yes)
                        .negativeText(android.R.string.no)
                        .onPositive { _, _ -> deleteChapters(getSelectedChapters()) }
                        .show()
            }
            else -> return false
        }
        return true
    }

    override fun onDestroyActionMode(mode: ActionMode) {
        adapter.mode = FlexibleAdapter.MODE_SINGLE
        adapter.clearSelection()
        actionMode = null
    }

    fun getSelectedChapters(): List<ChapterItem> {
        return adapter.selectedPositions.map { adapter.getItem(it) }
    }

    fun destroyActionModeIfNeeded() {
        actionMode?.finish()
    }

    fun selectAll() {
        adapter.selectAll()
        setContextTitle(adapter.selectedItemCount)
    }

    fun markAsRead(chapters: List<ChapterItem>) {
        presenter.markChaptersRead(chapters, true)
        if (presenter.preferences.removeAfterMarkedAsRead()) {
            deleteChapters(chapters)
        }
    }

    fun markAsUnread(chapters: List<ChapterItem>) {
        presenter.markChaptersRead(chapters, false)
    }

    fun markPreviousAsRead(chapter: ChapterItem) {
        val chapters = if (presenter.sortDescending()) adapter.items.reversed() else adapter.items
        val chapterPos = chapters.indexOf(chapter)
        if (chapterPos != -1) {
            presenter.markChaptersRead(chapters.take(chapterPos), true)
        }
    }

    fun downloadChapters(chapters: List<ChapterItem>) {
        destroyActionModeIfNeeded()
        presenter.downloadChapters(chapters)
        if (!presenter.manga.favorite){
            recycler.snack(getString(R.string.snack_add_to_library), Snackbar.LENGTH_INDEFINITE) {
                setAction(R.string.action_add) {
                    presenter.addToLibrary()
                }
            }
        }
    }

    fun bookmarkChapters(chapters: List<ChapterItem>, bookmarked: Boolean) {
        destroyActionModeIfNeeded()
        presenter.bookmarkChapters(chapters, bookmarked)
    }

    fun deleteChapters(chapters: List<ChapterItem>) {
        destroyActionModeIfNeeded()
        DeletingChaptersDialog().show(childFragmentManager, DeletingChaptersDialog.TAG)
        presenter.deleteChapters(chapters)
    }

    fun onChaptersDeleted() {
        dismissDeletingDialog()
        adapter.notifyItemRangeChanged(0, adapter.itemCount)
    }

    fun onChaptersDeletedError(error: Throwable) {
        dismissDeletingDialog()
        Timber.e(error)
    }

    fun dismissDeletingDialog() {
        (childFragmentManager.findFragmentByTag(DeletingChaptersDialog.TAG) as? DialogFragment)
                ?.dismissAllowingStateLoss()
    }

    override fun onItemClick(position: Int): Boolean {
        val item = adapter.getItem(position) ?: return false
        if (actionMode != null && adapter.mode == FlexibleAdapter.MODE_MULTI) {
            toggleSelection(position)
            return true
        } else {
            openChapter(item.chapter)
            return false
        }
    }

    override fun onItemLongClick(position: Int) {
        if (actionMode == null)
            actionMode = (activity as AppCompatActivity).startSupportActionMode(this)

        toggleSelection(position)
    }

    fun onItemMenuClick(position: Int, item: MenuItem) {
        val chapter = adapter.getItem(position)?.let { listOf(it) } ?: return

        when (item.itemId) {
            R.id.action_download -> downloadChapters(chapter)
            R.id.action_bookmark -> bookmarkChapters(chapter, true)
            R.id.action_remove_bookmark -> bookmarkChapters(chapter, false)
            R.id.action_delete -> deleteChapters(chapter)
            R.id.action_mark_as_read -> markAsRead(chapter)
            R.id.action_mark_as_unread -> markAsUnread(chapter)
            R.id.action_mark_previous_as_read -> markPreviousAsRead(chapter[0])
        }
    }

    private fun toggleSelection(position: Int) {
        adapter.toggleSelection(position)

        val count = adapter.selectedItemCount
        if (count == 0) {
            actionMode?.finish()
        } else {
            setContextTitle(count)
            actionMode?.invalidate()
        }
    }

    private fun setContextTitle(count: Int) {
        actionMode?.title = getString(R.string.label_selected, count)
    }
}
